### DISCUSSION TITLE
Getting a list of authorized clients for a subject

### DISCUSSION THREAD

**Discussion by adamcooke** [ory/hydra#2938] [Q&A] [answered]
Hi there.

I'm looking to provide my users with a list of clients that they have previously authorised and thus allow them to revoke any outstanding consents.

I can see the `GET /oauth2/auth/sessions/consent` endpoint but this seems to only include clients where the duration set for `remember` has not expired. The consent session seems to be able to expire while still allowing access tokens to continue working. 

If I was to use this to list authorised clients to a user, there could well be a client which still has an active access token which is not listed by this endpoint.

Am I missing something rather obvious here? 

Many thanks!

**Comment by vinckr** [comment]
>I'm looking to provide my users with a list of clients that they have previously authorised and thus allow them to revoke any outstanding consents.

A very reasonable use case!

>If I was to use this to list authorised clients to a user, there could well be a client which still has an active access token which is not listed by this endpoint.

Please correct me if I am wrong, but if the consent session is expired, users would need to do the consent flow again to be able to login. 
The problem with access tokens staying active after revocation of consent is inherent to the OAuth2.0 protocol, that is why we do not recommend it for first-party scenarios, read more here: [Why you probably do not need OAuth2 / OpenID Connect](https://www.ory.sh/oauth2-openid-connect-do-you-need-use-cases-examples/).
If a user revokes a consent session, the connected access tokens are still active for their lifetime. The way to solve this is to keep a "block list" AFAIK. 

From oauth.com

> This means there is no way to expire those tokens directly, so instead, the tokens are issued with a short expiration time so that the application is forced to continually refresh them, giving the service a chance to revoke an application’s access if needed.

**Reply by adamcooke** [reply]
> Please correct me if I am wrong, but if the consent session is expired, users would need to do the consent flow again to be able to login.

If the client goes through the authorisation flow again, yes, if a consent session has expired then the user will need to "Accept" the scopes and the app will be issued with a new access token.

> If a user revokes a consent session, the connected access tokens are still active for their lifetime. 

That's correct but also includes refresh tokens I believe, which means that an app could continue to have access to APIs for a considerable period after consent has been withdrawn. Hydra will refresh access tokens without consulting with anyone and even if there are no valid consent sessions.

**Comment by pflipp** [comment]
as a resource owner  i just want to know which apps currently has access to my resources and i want to be able revoke these "permissions" from an app.

i think currently this isn't possible as i would need the possibility to list all clients with valid refresh/access tokens for a subject and the possibility to revoke all tokens for a client and subject.

for me this has nothing to do with session termination/revocation.

maybe off-topic but:
> If a user revokes a consent session, the connected access tokens are still active for their lifetime.

the [API-Reference - revokeConsentSessions](https://www.ory.sh/hydra/docs/reference/api#operation/revokeConsentSessions) says

> [...] and invalidates all associated OAuth 2.0 Access Tokens.

**Comment by aeneasr** [comment, marked as answer]
If this is currently not possible (seeing a list of all clients which still have authorization) I think we need to introduce this feature. Please feel free to create an issue and a reproducible case or example API requests & responses of the way things look now and how you would expect them to look.

**Reply by mig5** [reply]
@aeneasr I have put together a proposed feature for a 'Manage existing consents' flow that would allow interacting with the DELETE and GET routes for listing consents in the Admin API, via the Login&Consent apps. The 'feature' is that a request to the public API would validate the id_token or access_token and and redirect to the Login&Consent app with challenge tokens for the Login&Consent app to use in those requests to the Admin API.

In short - just another 'flow' like Login, Consent and Logout, but for 'Managing existing consents'

https://github.com/ory/hydra/issues/3134

**Reply by vinckr** [reply]
Thanks for that request and taking the time to think this through! We will take a look as soon as we have some bandwidth and comment in the issue 🙏 .
